### PR TITLE
[FLINK-16160][table-common] Support computed column, proctime and watermark in Schema descriptor

### PR DESCRIPTION
**Pull request by docete**
## What is the purpose of the change

In ConnectTableDescriptor#createTemporaryTable, the proctime/rowtime properties are ignored so the generated catalog table is not correct. This PR fix string-based representation of TableSchema and support computed column, proctime and watermark in Schema descriptor.

## Brief change log

- a3f05f8 Fix misuse of prefix of connector-related properties
- 1ef55e3 Port CustomConnectorDescriptor to flink-table-api-java
- acc308a Fix string-based representation of TableSchema and support computed column, proctime and watermark in Schema descriptor.

## Verifying this change

This change is covered by tests, such as *TimeAttributeITCase*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Schema.java [hunk 1]
```diff
@@ -183,11 +250,26 @@ public Schema proctime() {
...
 	}
 
+	/**
+	 * Specifies the computed column as a processing-time attribute.
+	 *
+	 * <p>E.g. proctime("pt")
+	 */
+	public Schema proctime(String proctimeAttribute) {
+		addField(proctimeAttribute, "TIMESTAMP(3)", "PROCTIME()");
+		tableSchema.get(proctimeAttribute).put(SCHEMA_PROCTIME, "true");
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Schema.java]: Is there any problems if we don't add this property?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/descriptors/CustomConnectorDescriptor.java [hunk 2]
```diff
@@ -16,18 +16,16 @@
...
-import org.apache.flink.table.descriptors.DescriptorProperties;
+import org.apache.flink.annotation.PublicEvolving;
 
 import java.util.Map;
 
 /**
  * Describes a custom connector to an other system.
  */
-@Internal
+@PublicEvolving
```
[the remark this answers is not in the document]
Comment (docete) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/descriptors/CustomConnectorDescriptor.java]: There are many user-defined table sources (and related factory). I think we should provide a  common Connector Descriptor. Or user must define an extra descriptor for their table source. That's tedious and not user friendly.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Schema.java [hunk 3]
```diff
@@ -85,11 +111,25 @@ public Schema schema(TableSchema schema) {
...
+	 * Adds a field with the field name and the data type. Required.
+	 * This method can be called multiple times. The call order of this method defines
+	 * also the order of the fields in a row.
+	 *
+	 * @param fieldName the field name
+	 * @param fieldType the type information of the field
+	 * @param fieldExpr Computed column expression, it should be a SQL-style expression whose
+	 *                  identifiers should be all quoted and expanded.
+	 */
+	public Schema field(String fieldName, DataType fieldType, String fieldExpr) {
```
[the remark this answers is not in the document]
Comment (docete) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Schema.java]: The basic of quoting expression and deriving data type is SqlNode in DDL code base. I'm afraid it's hard achieve that in descriptor API. Any ideas?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Schema.java (hunk 1 above)
Comment (docete) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Schema.java]: The old `field(...).proctime()` api use this to identify a proctime field. We add this for back-compatibility. And we also use this property to deserialize/serialize TableSchema in DescriptorProperties.

**Pull request closed without merging**